apps/aecore/src/aec_db_gc.erl

```diff
@@ -0,0 +1,211 @@
+
+%% %%%===================================================================
+%% %%% API
+%% %%%===================================================================
+
+start_link() ->
+    #{enabled := Enabled, interval := Interval, history := History} = config(),
+    start_link(Enabled, Interval, History).
+
+start_link(Enabled, Interval, History) ->
+    gen_statem:start_link({local, ?MODULE}, ?MODULE, [Enabled, Interval, History], []).
+
+
+-ifdef(EUNIT).
+maybe_garbage_collect() -> nop.
+-else.
+%% this should be called when there are no processes modifying the block state
+%% (e.g. aec_conductor on specific places)
+maybe_garbage_collect() ->
+    gen_statem:call(?MODULE, maybe_garbage_collect).
+-endif.
+
+stop() ->
+    gen_statem:stop(?MODULE).
+
+%%%===================================================================
+%%% gen_statem callbacks
+%%%===================================================================
+
+init([true, Interval, History]) when Interval =< History ->
+    lager:error("GC interval ~p must be greater than history ~p", [Interval, History]),
+    {stop, interval_too_short};
+init([Enabled, Interval, History]) ->
+    if Enabled ->
+            aec_events:subscribe(top_changed),
+            aec_events:subscribe(chain_sync);
+       true ->
+            ok
+    end,
+    Data = #data{enabled  = Enabled,
+                 interval = Interval,
+                 history  = History,
+                 synced   = false,
+                 height   = undefined,
+                 hashes   = undefined},
+    {ok, idle, Data}.
+
+
+%% once the chain is synced, there's no way to "unsync"
+handle_event(info, {_, chain_sync, #{info := {chain_sync_done, _}}}, idle,
+             #data{enabled = true} = Data) ->
+    aec_events:unsubscribe(chain_sync),
+    {keep_state, Data#data{synced = true}};
+handle_event(info, {_, top_changed, #{info := #{height := Height}}}, idle,
+             #data{interval = Interval, history = History,
+                   enabled = true, synced = true,
+                   height = undefined, hashes = undefined} = Data)
+  when Height rem Interval == 0 ->
+    Top = self(),
+    Pid = spawn_link(
+            fun () ->
+                    FromHeight = max(Height - History, 0),
+                    {Time, {ok, Hashes}} = ?TIMED(collect_reachable_hashes(FromHeight, Height)),
+                    ets:give_away(Hashes, Top, {{FromHeight, Height}, Time})
+            end),
+    {keep_state, Data#data{height = Height, hashes = Pid}};
+handle_event(info, {'ETS-TRANSFER', Hashes, _, {{FromHeight, ToHeight}, Time}}, idle,
+             #data{enabled = true, hashes = Pid} = Data)
+  when is_pid(Pid) ->
+    ?LOG("scanning of ~p reachable hashes in range <~p, ~p> took ~p seconds",
+         [ets:info(Hashes, size), FromHeight, ToHeight, Time / 1000000]),
+    {next_state, ready, Data#data{hashes = Hashes}};
+
+handle_event(info, {_, top_changed, #{info := #{block_type := key, height := Height}}}, ready,
```

Comment:
I suggest that the `top_changed` events are either handled subsequently (not separated by another event), or handled in one callback and distinquished with a `case`.

Reply:
Normally I would do it that way, but here the flow is more natural - after top_changed (under specific circumstances), the ets transfer happens, and after that, we process top_changed differently. So it's nice top to bottom reading...